SPEC.md

```diff
@@ -109,13 +109,15 @@ A network configuration consists of a JSON object with the following keys:
 
 - `cniVersion` (string): [Semantic Version 2.0](https://semver.org) of CNI specification to which this configuration list and all the individual configurations conform. Currently "1.1.0"
 - `cniVersions` (string list): List of all CNI versions which this configuration supports. See [version selection](#version-selection) below.
-- `name` (string): Network name. This should be unique across all network configurations on a host (or other administrative domain).  Must start with an alphanumeric character, optionally followed by any combination of one or more alphanumeric characters, underscore, dot (.) or hyphen (-).
+- `name` (string): Network name. This should be unique across all network configurations on a host (or other administrative domain).  Must start with an alphanumeric character, optionally followed by any combination of one or more alphanumeric characters, underscore, dot (.) or hyphen (-). Must not contain characters disallowed in file paths. A path segment (such as a filesystem directory) with the same name as the network name, containing one or more plugin configuration JSON objects for that network, should exist at the same level as the network configuration object itself.
 - `disableCheck` (boolean): Either `true` or `false`.  If `disableCheck` is `true`, runtimes must not call `CHECK` for this network configuration list.  This allows an administrator to prevent `CHECK`ing where a combination of plugins is known to return spurious errors.
-- `plugins` (list): A list of CNI plugins and their configuration, which is a list of plugin configuration objects.
+- `disableLoadPluginsFromPath` (boolean): Either `true` or `false`. If `false` (default), indicates [plugin configuration objects](#plugin-configuration-objects) should be loaded from a sibling directory with the same name as the network `name` field. These sibling directories should exist at the same path as the network configuration itself. Any valid plugin configuration objects within a named sibling directory will be appended to the final list of `plugins` for that network name. If set to `true`, plugin configuration objects in sibling directories will be ignored. If `plugins` is not present in the network configuration, `disableLoadPluginsFromPath` cannot be set to `true`.
```

Comment:
I've been thinking about this a lot, and I'm of two minds as to whether or not we should allow configuring this or not.

This is not a network-specific option, rather a something the administrator would choose to do. That is to say, administrators should say (via ContainerD's configuration) whether or not they would like to disable drop-ins.

Given that this expresses administrative intent, and the whole purpose of this exercise is to allow for a cleaner meshing between vendor-supplied configurations, maybe we get rid of this option altogether? Thoughts?

Reply:
> Given that this expresses administrative intent, and the whole purpose of this exercise is to allow for a cleaner meshing between vendor-supplied configurations, maybe we get rid of this option altogether? Thoughts?

It is mostly here for API compatibility with runtimes.

If we don't have this flag in the network config, we have to alter `libcni` config file loading APIs more drastically and backwards compatibility between network configs with inline plugins and network configs with subdirectory plugins gets pushed entirely to the container runtimes to manage.

Which feels a bit gross since it's our library, and our config, and our spec.

Putting the flag in the netconfig like this makes the back compat story really straightforward and will make integrating this into container runtimes pretty trivial, even if they don't change the libcni APIs they use, so for those reasons I'd prefer to keep it.


That being said, since it is an opt-in disable, a future spec revision could (and arguably should) drop it, which would force us to deprecate some of our libcni APIs and make container runtimes load netconfs and plugin confs themselves, separately and directly (they can do this today as well, if they choose to use the `NetworkPluginConfFromBytes` APIs added here)

Reply:
In light of the other spec changes tho, if you want to make this `disableLoadPlugins` or `disablePluginAggregation`, open to that.

Reply:
Consensus here was: https://hackmd.io/jU7dQ49dQ86ugrXBx1De9w?both#2024-04-08

1. Keep opt-out in config
2. Rename this value to be loading-source-agnostic